[WIP] Basic structure for Joint probability classes.

akash9712 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
Wrote the classes to implement multivariate distributions. These include `JointPSpace`, `JointDistributions` and `MarginalPSpace`. Classes for predefined distributions are implemented in joint_rv_types.py.

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed


#### Other comments

akash9712 committed:
[WIP] Basic structure for Joint probability classes.

Wrote the classes to implement joint probability, for two cases:
1) All the parameters given to `Joint` are independent random symbols
2) At least one of the parameters are symbols being used as random
variables for the joint distribution.

akash9712 committed:
Made minor changes and added documentation for `Joint`

Made changes to `density` in rv.py, added `__call__` for JointDistribution.
Also added documentation for `Joint`.

jksuom commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,195 @@
+        _set = rv_sets[0]
+        for i in rv_sets[1:]:
+            _set = _set*i
+        return _set
```
Would it be useful to store the result to avoid repeated computations?

akash9712 committed:
Added import statements and made minor corrections in joint_rv.py

Upabjojr commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,195 @@
+from sympy.sets.contains import Contains
+from sympy.integrals.integrals import Integral, integrate
+
+class JointDomain(RandomDomain):
```
do we need this class? Aren't joint domains and product domains the same thing?

Upabjojr commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,195 @@
+    def __contains__(self, other):
+        return other in self.set()
+
+class JointPSpace(PSpace):
```
Should this be a subclass of `ProductPSpace`?

akash9712 replied:
Yes, I'll add it as a subclass.

Upabjojr commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,195 @@
+class JointPSpace(PSpace):
+    """
+
+    """
```
We need a special constructor for this class.

akash9712 committed:
Removed joint_rv imports from stats/__init__.py

akash9712 committed:
Changes in joint_rv.py

Removed `JointDomain` and added `ProductPSpace` as parent class for
`JointPSpace`

akash9712 committed:
Allow given conditions of type equality for `Probability`

If the `given_condition` while calculating probability of an event were
of the type `Eq`, an error was raised. This is now fixed. Tests were
added.

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into jointprob

akash9712 committed:
Added multivariate normal distribution.

Density for multivariate normal distribution can now be calculated.

Upabjojr commented on sympy/stats/rv.py:
```diff
@@ -742,7 +758,7 @@ def doit(self, evaluate=True, **kwargs):
         if (isinstance(expr, RandomSymbol) and
             hasattr(expr.pspace, 'distribution') and
-            isinstance(pspace(expr), SinglePSpace)):
+            isinstance(pspace(expr), (SinglePSpace, JointPSpace))):
```
superclass?

akash9712 replied:
`JointPSpace` inherits from `ProductPSpace`, which has to be used if it is not an instance of either `SinglePSpace` or `JointPSpace`. These both inherit from `PSpace`, but `ProductPSpace` also inherits from it, so it would be an issue.

Upabjojr replied:
maybe a common superclass `PSpaceWithDistribution`?

akash9712 replied:
@Upabjojr With the current implementation of `SinglePSpace` and `JointPSpace`, the only method the superclass would have will be `distribution`. I don't the a super class is needed.

akash9712 commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,170 @@
+    """
+
+    """
+    def __new__(cls, name, syms, dist, **args):
```
`name` is same as the symbol used to initialize the object, same as what is done for other random variables. Eg: In `Y = Poisson('Y', 1)`, 'Y' is analogous to `name`.

Upabjojr replied:
`RandomSymbol` supports name. I don't think the product space class should have a name.

akash9712 replied:
`SinglePSpace` takes a symbol as one of the arguments, which is the same that is used for the initialisation of the random symbol. Should the same not be done for the joint space?

jksuom replied:
It seems to me that a joint space does not have a single distinguished random variable whose name would describe the whole space unlike `SinglePSpace`.

Upabjojr replied:
I believe the symbol taken as argument to `SinglePSpace` is needed for independence (in case of more pspaces with the same distribution). You are already passing `syms`, right? Aren't `syms` already identifying the joint probability space?

akash9712 replied:
Oh okay. I'll make the correction.

akash9712 committed:
Added marginal distribution.

Added marginal distribution for Joint probability spaces.

akash9712 committed:
Added multivariate distribution types.

Added multivariate Laplace, multivariate T, and normal-gamma
disributions.

akash9712 committed:
Resolved test failures in core/tests/test_args.py

Added tests for missing classes and modified class constructors in
joint_rv.py joint_rv_types.py.

Upabjojr commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,176 @@
+        raise NotImplementedError()
+
+
+class MarginalPSpace(JointPSpace):
```
what's the advantage of having this class separate from `JointPSpace`?

akash9712 replied:
Having it as a seperate class allows a different constructor, and each marginal probability space can  have a joint space associated with it. Also, creating a `JointPSpace` requires to specify a distribution, which is not needed for `MarginalPSpace`, which is identified on the basis of the joint probability space from which it is derived and the symbols in it.

akash9712 replied:
The marginal pspace object is not being returned anywhere right now. It's only being used internally to calculate the density. Should I add a function that allows the user to create and use the marginal probability space similar to a joint probability space?

Upabjojr replied:
I'm a bit skeptical about inheriting from `ProductPSpace`: the way classes are designed in SymPy requires the arguments of a subclass constructor to match the arguments of the superclass constructor. You add `syms` as first arguments, which disrupts this logic.

Upabjojr replied:
what about making `ProductPSpace` an abstract class and move the constructor to another subclass, like `IndependentProductPSpace` (this would manage the class for independent prob spaces)?

Upabjojr replied:
>  Also, creating a JointPSpace requires specifying a distribution, which is not needed for MarginalPSpace; it can be identified on the basis of the joint probability space from which it is derived and the symbols in it.

If you change the constructor, then you should not inherit from `JointPSpace`. The marginal pspace is technically still a joint pspace. I see the reason for a separate class in the case you cannot calculate the marginal distribution, so you need to support the unevaluated expression, but I wouldn't spend too much time on this in this PR.

akash9712 replied:
Okay, I'll add a method for calculating the marginal density in `JointPSpace` and remove `MarginalPSpace` for now.

Upabjojr commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,176 @@
+        NamedArgsMixin, ProductDomain)
+from sympy.core.containers import Tuple
+
+class JointPSpace(ProductPSpace):
```
There is a logical problem here: `ProductPSpace` is implemented for the category of product spaces of independent random variables. If you inherit `ProductPSpace`, you are meant to deal with a subset of products of independent pspaces. Unfortunately the joint pspaces are a separate category. This current class inheritance structure is not good.

Furthermore, SymPy convention requires the subclass to have the same arguments in the constructor as the superclass.

Upabjojr replied:
One solution could be to make `ProductPSpace` abstract, and move the implementation of the independent pspaces to another subclass, like `IndependentProductPSpace`.

akash9712 replied:
Sorry, I wasn't aware of the convention. I will implement the abstract class as suggested.
Will this need to undergo the deprecation process?

Upabjojr replied:
no, because the new version of SymPy is still not out. Furthermore, if the API is hidden (cannot be imported directly), we can break it.

Upabjojr replied:
That's why I suggest to keep the API hidden until the end of the summer of code. Unhiding only takes 5 minutes.

akash9712 replied:
`JointPSpace` could also be made seperate from `ProductPSpace` altogether, by inheriting directly from `PSpace`

akash9712 committed:
Removed `MarginalPSpace`, made `ProductPSpace` abstract

Removed `MarginalPSpace` class and added method `marginal_distribution`
to `JointPSpace` to calculate marginal distribution function.
Also, renamed `ProductPSpace` to `IndpendentProductPSpace` and made
`ProductPSpace` as an abstract parent class for `JointPSpace` and
`IndependentProductPSpace`. Added tests for the above changes.

akash9712 committed:
Added test for `marginal_density` method of `NormalGammaDistribution`

akash9712 commented:
@Upabjojr I have made the changes according to your comments. Please review.

akash9712 committed:
Added newline at end of joint_rv.py and test_joint_rv.py

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,247 @@
+
+    A random symbol
+    """
+    return rv(MultivariateNormalDistribution, syms, mu, sigma)
```
I believe this should return an instance of `RandomSymbol`

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,247 @@
+                (_mu - sym)))[0])
+
+
+def MultivariateNormal(syms, mu, sigma):
```
I'm not sure about having more symbols defining a multivariate normal. The multivariate normal should still be represented by a single symbol.

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,247 @@
+
+    A random symbol
+    """
+    return rv(MultivariateLaplaceDistribution, syms, mu, sigma)
```
same reasoning

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,247 @@
+
+    A random symbol
+    """
+    return rv(MultivariateTDistribution, syms, mu, sigma, v)
```
same problem

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,247 @@
+
+    A random symbol
+    """
+    return rv(NormalGammaDistribution, syms, mu, lamda, alpha, beta)
```
same problem

akash9712 committed:
Modified `JointPSpace`.

Modified `JointPSpace` to take only one symbol as input for the joint
probability space, and return a random symbol instead of the whole
space. Also, allow indexing on the random variable. Marginal density can
be calculated using either only the component index number, or the
indexed random variable. Added test, and modified doctests in
joint_rv.py accordingly.

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,248 @@
+                (_mu - sym)))[0])
+
+
+def MultivariateNormal(syms, mu, sigma):
```
aren't we going to mix this with `Normal`? do you think we should keep this separate from `Normal`?

akash9712 replied:
Sorry, I forgot about that. I'll mix the two. Should I remove this, or just add the extra functionality to `Normal`?

Upabjojr replied:
only `Normal`. If the parameters are matrices or lists, it should create a multivariate random symbol.

Upabjojr commented on sympy/stats/rv.py:
```diff
@@ -267,8 +267,19 @@ def _hashable_content(self):
     def free_symbols(self):
         return {self}
 
-
```
we usually prefer two newlines on the root indentation

akash9712 commented on sympy/stats/joint_rv.py:
```diff
@@ -0,0 +1,165 @@
+    def probability(self, condition):
+        raise NotImplementedError()
+
+class JointDistribution(Basic, NamedArgsMixin):
```
I think it would be better this way, in case we have to add a distribution that has both discrete and continuous components.

Upabjojr replied:
OK

akash9712 committed:
Removed `MultivariateNormal`, added the functionality to `Normal`.

akash9712 commented:
Ping @Upabjojr

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -1974,13 +1977,30 @@ def Normal(name, mean, std):
     .. [2] http://mathworld.wolfram.com/NormalDistributionFunction.html
     """
 
+    if isinstance(mean, (list)) and isinstance(std, list):
```
needs to check against matrices as well, I believe there's a common class, `MatrixBase` I think it's the name.

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,228 @@
+# 'NormalGamma'
+# ]
+
+def rv(cls, sym, *args):
```
maybe let's call it `multivariate_rv`

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,228 @@
+        return gamma((k + v)/2)/(gamma(v/2)*(v*pi)**(k/2)*sqrt(det(sigma)))\
+        *(1 + 1/v*((args - mu).transpose()*sigma_inv*(args - mu))[0])**((-v - k)/2)
+
+def MultivariateT(syms, mu, sigma, v):
```
maybe code for another PR

akash9712 replied:
Made the changes. I'll send the seperate PR after this is merged.

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,228 @@
+#-------------------------------------------------------------------------------
+# Multivariate T-distribution ---------------------------------------------------------
+
+class MultivariateTDistribution(JointDistribution):
```
separate PR

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,228 @@
+            *((args.transpose()*sigma_inv*args)[0])))\
+        *exp((args.transpose()*sigma_inv*mu)[0])
+
+def MultivariateLaplace(syms, mu, sigma):
```
separate PR

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -0,0 +1,228 @@
+#-------------------------------------------------------------------------------
+# Multivariate laplace distribution ---------------------------------------------------------
+
+class MultivariateLaplaceDistribution(JointDistribution):
```
separate PR

akash9712 committed:
Removed joint distributions and added check in `Normal`

Removed some joint probability distributions, added check for matrices
in `Normal` for arguments, and modified tests for the same.

akash9712 commented:
Ping @Upabjojr 
I have made the changes as suggested.

Upabjojr commented:
OK, merged
